### PR TITLE
Workarounds for CUDA 12.1 and 12.2

### PR DESCRIPTION
**Pull request by havogt**
CUDA 12.1 and 12.2 have a problem with constexpr, e.g. in the context of CTAD, see #1766. The workaround is to do pre-C++17 `make_tuple`-construction or construct from a (possibly moved-from) lvalue.

CI: add GCC + CUDA 12.1/12.2 and NVHPC 23.7

### CODE REVIEW

**Comment by havogt** [comment]
launch jenkins

File: tests/include/nvcc_workarounds.hpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+#include <gridtools/common/tuple.hpp>
+
+namespace gridtools {
+    namespace nvcc_workarounds {
+
+        // see https://github.com/GridTools/gridtools/issues/1766
+        template <class T>
+        constexpr auto make_1_tuple(T &&t) {
+            return tuple(t);
```
Comment (havogt) [inline tests/include/nvcc_workarounds.hpp:19]:
```suggestion
            return tuple(std::forward<T>(t));
```

File: tests/regression/fn/fn_cartesian_vertical_advection.cpp [hunk 2]
```diff
@@ -27,7 +28,7 @@ namespace {
 
     struct u_forward_scan : fwd {
         static GT_FUNCTION constexpr auto prologue() {
-            return tuple(scan_pass(
+            return nvcc_workarounds::make_1_tuple(scan_pass(
```
Comment (havogt) [inline tests/regression/fn/fn_cartesian_vertical_advection.cpp:31]: in tests, our strategy is usually to not #ifdef-enable the workarounds, but just keep it for all compilers.

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch perftests

File: include/gridtools/fn/column_stage.hpp [hunk 3]
```diff
@@ -101,5 +101,13 @@ namespace gridtools::fn {
     using column_stage_impl_::column_stage;
     using column_stage_impl_::fwd;
     using column_stage_impl_::merged_column_stage;
+#if defined(__NVCC__) && (__CUDACC_VER_MAJOR__ == 12 && __CUDACC_VER_MINOR__ >= 1 && __CUDACC_VER_MINOR__ <= 2)
```
Comment (petiaccja) [inline include/gridtools/fn/column_stage.hpp:104]: You could extract this expression and use it in al 3 places in this PR:
```suggestion
#if GRIDTOOLS_CUDA_CTAD_WORKAROUND
```

File: include/gridtools/fn/column_stage.hpp [hunk 4]
```diff
@@ -101,5 +101,13 @@ namespace gridtools::fn {
...
+    // workaround CTAD issue in CUDA 12.1, 12.2 (https://github.com/GridTools/gridtools/issues/1766)
```
Comment (petiaccja) [inline include/gridtools/fn/column_stage.hpp:105]: The comment is good, but if you declare the workaround condition as a "variable" then no need for it.

File: tests/include/nvcc_workarounds.hpp [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2021, ETH Zurich
```
Comment (petiaccja) [inline tests/include/nvcc_workarounds.hpp:4]:
```suggestion
 * Copyright (c) 2014-2023, ETH Zurich
```

File: tests/unit_tests/common/test_tuple.cpp [hunk 6]
```diff
@@ -331,5 +331,13 @@ namespace gridtools {
             EXPECT_EQ(a, 42);
             EXPECT_EQ(b, 2.5);
         }
+
+#if defined(__CUDACC_VER_MAJOR__) && __CUDACC_VER_MAJOR__ == 12 && __CUDACC_VER_MINOR__ >= 1 && __CUDACC_VER_MINOR <= 2
+#else
```
Comment (petiaccja) [inline tests/unit_tests/common/test_tuple.cpp:336]:
```suggestion
#if !GRIDTOOLS_CUDA_WORKAROUND
```

**Review by petiaccja — CHANGES_REQUESTED** [review]
Some minor cleanups, otherwise not much to do in my opinion.

One question though, why not just use `gridtools::make_tuple` everywhere, why a `make_1_tuple`?

**Comment by havogt** [comment]
> Some minor cleanups, otherwise not much to do in my opinion.
> 
> One question though, why not just use `gridtools::make_tuple` everywhere, why a `make_1_tuple`?

I wanted to emphasize that this is a workaround that might be needed for users. If you see weird behavior in your code, you might compare against the example and find the problem.

File: tests/include/nvcc_workarounds.hpp (hunk 5 above)
Comment (havogt) [reply tests/include/nvcc_workarounds.hpp:4]: will update in one go in a next pr

File: include/gridtools/fn/column_stage.hpp (hunk 3 above)
Comment (havogt) [reply include/gridtools/fn/column_stage.hpp]: done

**Comment by havogt** [comment]
launch jenkins

File: tests/unit_tests/common/test_tuple.cpp [hunk 7]
```diff
@@ -331,5 +332,12 @@ namespace gridtools {
             EXPECT_EQ(a, 42);
             EXPECT_EQ(b, 2.5);
         }
+
+#if not GT_NVCC_WORKAROUND_1766
```
Comment (petiaccja) [inline tests/unit_tests/common/test_tuple.cpp:336]: I think it's best to use the special character operators for consistency:
```suggestion
#if !GT_NVCC_WORKAROUND_1766
```

**Review by petiaccja — CHANGES_REQUESTED** [review]

**Comment by petiaccja** [comment]
> > Some minor cleanups, otherwise not much to do in my opinion.
> > One question though, why not just use `gridtools::make_tuple` everywhere, why a `make_1_tuple`?
> 
> I wanted to emphasize that this is a workaround that might be needed for users. If you see weird behavior in your code, you might compare against the example and find the problem.

I understand, however, is this workaround ever gonna go away? (Ever = within 1-2 years.) The problem is, unless they fix their older compiler releases or we simply stop supporting those specific releases, we have to keep this workaround in the code, and then I'd rather use a proper interface that works consistently even if it's not obvious why that's used.

**Comment by havogt** [comment]
launch jenkins

**Review by petiaccja — APPROVED** [review]
As discussed, we'll keep the workaround function `make_1_tuple` because it's only used in the tests, so it's not visible to users. Users should be discouraged to use affected CUDA versions unless nVidia fixes the problem.

**Pull request merged**